src/type_checker.cpp

```diff
@@ -475,6 +486,27 @@ void TypeChecker::Visit(PostfixArithExprNode& postfix_expr) {
   postfix_expr.type = postfix_expr.operand->type->Clone();
 }
 
+void TypeChecker::Visit(RecordMemExprNode& mem_expr) {
+  mem_expr.expr->Accept(*this);
+  // id_expr->id is a record variable id.
+  const auto* id_expr = dynamic_cast<IdExprNode*>((mem_expr.expr).get());
+  assert(id_expr);
```

Comment:
> Is it a _todo_ for the `expr` to not be a record variable id? For example:
> 
> ```c++
> struct birth bd[3] = {...};
> bd[0].date;
> ```
> 
> Additionally, the constraints for structure and union members specified in 6.5.2.3 are:
> 
> 1. The first operand of the `.` operator must have a structure or union type, and the second operand must name a member of that type.
> 2. The first operand of the `->` operator must be a pointer to a structure or union, and the second operand must name a member of the type pointed to.
> 
> I believe we're missing the check on the type of the left-hand side. Also, we don't need to check if the symbol is declared when it's an `IdExprNode`, as this check is already performed in the visit function of `IdExprNode`. I noticed we are performing this check in the `FuncCallExprNode` as well, which should be addressed in a future PR.

Ｍake sense! I'll remove this id check and update it with type check.

Reply:
> Is it a _todo_ for the `expr` to not be a record variable id? For example:
> 
> ```c++
> struct birth bd[3] = {...};
> bd[0].date;
> ```
> 
> Additionally, the constraints for structure and union members specified in 6.5.2.3 are:
> 
> 1. The first operand of the `.` operator must have a structure or union type, and the second operand must name a member of that type.
> 2. The first operand of the `->` operator must be a pointer to a structure or union, and the second operand must name a member of the type pointed to.
> 
> I believe we're missing the check on the type of the left-hand side. Also, we don't need to check if the symbol is declared when it's an `IdExprNode`, as this check is already performed in the visit function of `IdExprNode`. I noticed we are performing this check in the `FuncCallExprNode` as well, which should be addressed in a future PR.

I've added an else statement since RecordType is either StructType or UnionType. Let me know what you think!

Reply:
I think we should check directly on the type of `mem_expr` instead of trying to cast it to the `IdExprNode` beforehand, which is because of the reason that I've shown above with an example.